### ISSUE TITLE
Per-layer checkboxes

### ISSUE THREAD

**Issue by Robinlovelace** [nptscot/nptscot.github.io#264]
Currently we have this:

![image](https://github.com/user-attachments/assets/f61b550f-cdab-45ab-a312-01b4374e3a22)

We need checkboxes for each category.

Please prioritise @mvl22. I imagine it's a quick fix.

We may want to enable this 'per category' checkboxes for other layers but let's leave it at that for this issue.

**Comment by Robinlovelace** [comment]
Analogous design can be found in screenshot below for reference:

![image](https://github.com/user-attachments/assets/aeaf8b5a-3f0f-4bb8-88c3-e7c4d45332ac)

Source: https://acteng.github.io/atip/browse.html?style=dataviz&cycle_paths=track%2Cshared_use_segregated#11.12/53.7649/-1.5613

**Comment by Robinlovelace** [comment]
Nice thing about the above: the URL contains info on which layers are showing, making it easier for users to share the specific information they are looking at / app state.

**Comment by mvl22** [comment]
Yes, it's just a filter. Will add shortly as a priority.

**Comment by Robinlovelace** [comment]
Any updates on this @mvl22? this is the top UI priority now I imagine.

**Comment by mvl22** [comment]
Yes, half way through this now.

**Comment by Robinlovelace** [comment]
Just checking-in on this as it was marked as "VH" (very high) importance by @anguscalder and team with this description:

> We REALLY need to be able to toggle on/off local access/non-arterial links with tick box!

**Comment by Robinlovelace** [comment]
> Yes, half way through this now.

@mvl22 should be close to complete right? Any input needed let us know.

**Comment by mvl22** [comment]
For cycle infrastructure layer, all to have checkboxes, but `Painted Cycle Lane` checkbox to be off by default.

**Comment by Robinlovelace** [comment]
Nice! Those checkboxes do look a bit small, maybe make the legend colour bar a bit wider?

**Comment by Robinlovelace** [comment]
Heads-up @mvl22 I've updated the title because I think we need per-layer checkboxes for all layers.

It would be very handy to be able to view just the "Do not use (non-compliant intervention)" segments, as shown below, for example:

![Image](https://github.com/user-attachments/assets/fe9cb4a0-71a5-40ef-86f4-a954e232b092)

Source: https://nptscot.github.io/#/clos/#11.77/55.9471/-2.9394

**Comment by mvl22** [comment]
> Heads-up @mvl22 I've updated the title because I think we need per-layer checkboxes for all layers.

Yes, I'm coding this to be generic to all, as there's no reason not to.

**Comment by mvl22** [comment]
Blocked on #408.

The basic mechanism on the branch works fine, and is working for the Coherent network. But anything involving sublayers, i.e. layers that are changed by a drop-down to select which sublayer is used, basically doesn't work as those sublayers cannot yet identify the relevant field in play. Fixing the layer/sublayer split should resolve this.

Cc: @Robinlovelace

**Comment by mvl22** [comment]
All the legacy inconsistencies in the layer definitions have now been removed in #408, so this is now unblocked.

**Comment by mvl22** [comment]
The checkboxes functionality needs matches for both the `match` (i.e. string match) value and `step` (i.e. a range) value cases.

**Comment by mvl22** [comment]
> Get layer definitions standardised and legend bugs all fixed

Now done in #408 and PR #418.

**Comment by Robinlovelace** [comment]
Great work and thanks for keeping up-to-date on this.

**Comment by mvl22** [comment]
> - [x] Get checkboxes mechanism working
> - [x] Add support for both the `match` (i.e. string match) value and `step` (i.e. a range) value cases.

Filtering is now working and generically available for all layers where checkboxes are created, even for range fields where a complex filter has to be constructed.

However, it will **not** yet work in the `rnet` layer as that has special-case handling in the UI for dynamically constructing the field being shown from the set of three drop-downs, that I need to support before this functionality can be enabled.

![Image](https://github.com/user-attachments/assets/fd7be142-80c7-4772-98c2-9f3d06f30d49)

**Comment by Robinlovelace** [comment]
Impressive! The checkboxes embedded in the legends look pretty good to me. 

Wider and more out-there idea: could it be worth generalising some of this logic at some point, e.g. as a npm package?

Example of modular codebase that was split-out from a previous one-off project: https://github.com/acteng/govuk-svelte

**Comment by mvl22** [comment]
> Retain state upon redraw, for enabling/disabling layer

State memory implemented in 227c3032d1812ef520fe3fba230de9eeee3adacb.

**Comment by mvl22** [comment]
> Wider and more out-there idea: could it be worth generalising some of this logic at some point, e.g. as a npm package?

The logic in `src/nptui.js` is entirely generic (apart from two very isolated hacks to deal with the usual rnet/rnet-simplified switching problem) - it doesn't really know anything about what the layers are, just that that they are defined.

**Comment by Robinlovelace** [comment]
Thanks for the update. Looking forward to giving it a spin.

**Comment by mvl22** [comment]
> Get things working for the special-case rnet layer, which uses a compound sublayer ID from multiple form widgets

Turns out we don't want checkbox filtering for this, because we already have more detailed filter sliders. The two can't be mixed.

Implemented in 4da7b911b7bb5eb31460281f9113bf075d3fe95b.

**Comment by mvl22** [comment]
PR in #419 to get this live.

Not yet in URL state but everything else done.
